ground/src/main/java/com/google/android/ground/persistence/sync/LocalMutationSyncWorker.kt

```diff
@@ -82,63 +83,55 @@ constructor(
     return pendingMutations + failedMutationsEligibleForRetry
   }
 
-  /**
-   * Groups mutations by user id, loads each user, applies mutations, and removes processed
-   * mutations.
-   *
-   * @return `true` if all mutations are applied successfully, else `false`
-   */
-  private suspend fun processMutations(allMutations: List<Mutation>): Boolean {
-    val mutationsByUserId = allMutations.groupBy { it.userId }
-    val userIds = mutationsByUserId.keys
-    var noErrors = true
-    for (userId in userIds) {
-      val mutations = mutationsByUserId[userId]
-      val user = getUser(userId)
-      if (mutations == null || user == null) {
-        continue
-      }
-      val result = processMutations(mutations, user)
-      if (!result) {
-        noErrors = false
-      }
-    }
-    return noErrors
-  }
-
   /**
    * Applies mutations to remote data store. Once successful, removes them from the local db.
    *
-   * @return `true` if the mutations were successfully synced with [RemoteDataStore].
+   * @return `success` if the mutations were successfully synced with [RemoteDataStore].
    */
-  private suspend fun processMutations(mutations: List<Mutation>, user: User): Boolean {
-    check(mutations.isNotEmpty()) { "List of mutations is empty" }
-
+  private suspend fun processMutations(mutations: List<Mutation>): Result {
+    if (mutations.isEmpty()) return success()
+    val user = getUserFromMutations(mutations) ?: return failure()
     return try {
       mutationRepository.markAsInProgress(mutations)
       remoteDataStore.applyMutations(mutations, user)
       mutationRepository.finalizePendingMutationsForMediaUpload(mutations)
-      true
+      success()
     } catch (t: Throwable) {
-      // Mark all mutations as having failed since the remote datastore only commits when all
-      // mutations have succeeded.
       Timber.d(t, "Local mutation sync failed")
-      val crashlytics = FirebaseCrashLogger()
-      crashlytics.setSelectedSurveyId(mutations.first().surveyId)
-      crashlytics.logException(t)
-      mutationRepository.markAsFailed(mutations, t)
-      false
+      handleMutationSyncFailed(mutations, t)
+      retry()
     }
   }
 
-  private suspend fun getUser(userId: String): User? {
-    val user = localUserStore.getUserOrNull(userId)
+  /**
+   * Returns the user associated with the mutations. If missing, removes the mutations from local
+   * storage as well.
+   */
+  private suspend fun getUserFromMutations(mutations: List<Mutation>): User? {
+    val userIds = mutations.map { it.userId }.toSet()
+
+    check(userIds.size == 1) { "Found mutations for multiple users" }
```

Comment:
Do we really want to crash the app in this case? Wouldn't we instead want to log something to Crashlytics and ignore?

Reply:
Actually, this really can never occur in practice, since multiple users can't create data with the same LOI id. 

Instead of this, we should be checking that all user IDs are equal to the id of the current user, ignoring ones that are not and reporting to Crashlytics without crashing the app.

Reply:
> Do we really want to crash the app in this case? Wouldn't we instead want to log something to Crashlytics and ignore?

The reason for crashing the app is because the state is never expected to occur. Ignoring the error might lead to other side-effects which may go unnoticed.

> Instead of this, we should be checking that all user IDs are equal to the id of the current user, ignoring ones that are not and reporting to Crashlytics without crashing the app.

I've left a TODO for adding the check. I'm not sure what would happen if the app is opened in the background and the worker tries to fetch the logged in user which gets initialized only when opening the app.

Reply:
> The reason for crashing the app is because the state is never expected to occur. Ignoring the error might lead to other side-effects which may go unnoticed.

I agree we would want to know about this, but I think ignoring the unexpected mutations and processing the expected ones wouldn't make the situation worse, but it might allow users to continue working, and minimize the amount of data stuck on the device.

> I've left a TODO for adding the check. I'm not sure what would happen if the app is opened in the background and the worker tries to fetch the logged in user which gets initialized only when opening the app.

The worker is only cued due to user action, which means the app had to be opened at some point in the past. Those credentials are serialized used by the worker to access Firebase.

Reply:
Added check for currently logged in user